solarforecastarbiter/reference_forecasts/models.py

```diff
@@ -77,23 +79,51 @@ def get_nwp_model(func):
 
 
 def _resample_using_cloud_cover(latitude, longitude, elevation,
-                                cloud_cover, air_temperature, wind_speed):
+                                cloud_cover, air_temperature, wind_speed,
+                                start, end, interval_label):
     """
     Calculate all irradiance components from cloud cover.
+
+    Cloud cover from GFS is an interval average with ending label.
+    Air temperature and wind speed are instantaneous values.
+    Intervals are 1, 3, or 6 hours in length.
+    To accurately convert from cloud cover to irradiance, we need to
+    interpolate this data to subhourly resolution because solar position
+    and PV power calculations assume instantaneous inputs at each time.
+
+    Parameters
+    ----------
+    latitude : float
+    longitude : float
+    elevation : float
+    cloud_cover : pd.Series
+    air_temperature : pd.Series
+    wind_speed : pd.Series
+    interval_label : str
+        beginning, ending, or instant
     """
     # Interpolate cloud cover, temp, and wind to higher temporal resolution
     # because solar position and PV power calculations assume instantaneous
     # inputs. Why 5 minutes? It's a round number that produces order 10 data
     # points per hour, so it's reasonable for hour average calculations.
-    interpolator = partial(forecast.interpolate, freq='5min')
-    cloud_cover, air_temperature, wind_speed = list(
-        map(interpolator, (cloud_cover, air_temperature, wind_speed)))
+    freq = '5min'
+    label = datamodel.CLOSED_MAPPING[interval_label]
+    cloud_cover = cloud_cover.resample(freq).bfill()
+    interpolator = partial(forecast.interpolate, freq=freq)
+    air_temperature, wind_speed = list(
+        map(interpolator, (air_temperature, wind_speed)))
+    start_adj, end_adj = adjust_start_end_for_interval_label(interval_label,
+                                                             start, end)
+    slicer = partial(forecast.slice_arg, start=start_adj, end=end_adj)
+    cloud_cover, air_temperature, wind_speed = [
```

Comment:
this is the copy pasta part. It isn't much, but important enough that having one shared function makes it easier to understand each model. I know sometimes you``slicer(interpolate(v))`` so ``_interpolate_slice`` can have two sets of args like:

```python
def _interpolate_slice(forecast, *, slice_only_args, interpolate_slice_args, interval_label, start, end, freq):
    start_adj, end_adj = adjust_start_end_for_interval_label(interval_label, start, end)
    args = list(slice_only_args) # maybe deepcopy?
    args += [forecast.interpolate(v, freq=freq) for v in interpolate_slice_args]
    return slice_args(args, start=start_adj, end=end_adj)
```